nifi-nar-bundles/nifi-py4j-bundle/nifi-py4j-bridge/src/main/java/org/apache/nifi/py4j/PythonProcess.java

```diff
@@ -0,0 +1,293 @@
+        final String pythonCmd = pythonCmdFile.getName();
+        final File pythonCommandFile = new File(virtualEnvHome, "bin/" + pythonCmd);
+        final String pythonCommand = pythonCommandFile.getAbsolutePath();
+
+        final File controllerPyFile = new File(pythonFrameworkDirectory, PYTHON_CONTROLLER_FILENAME);
+        final ProcessBuilder processBuilder = new ProcessBuilder(pythonCommand, controllerPyFile.getAbsolutePath());
+        processBuilder.environment().put("JAVA_PORT", String.valueOf(listeningPort));
+        processBuilder.environment().put("LOGS_DIR", pythonLogsDirectory.getAbsolutePath());
+        processBuilder.environment().put("ENV_HOME", virtualEnvHome.getAbsolutePath());
+        processBuilder.environment().put("PYTHONPATH", pythonApiDirectory.getAbsolutePath());
+        processBuilder.environment().put("PYTHON_CMD", pythonCommandFile.getAbsolutePath());
+        processBuilder.inheritIO();
+
+        logger.info("Launching Python Process {} {} with working directory {} to communicate with Java on Port {}",
+            pythonCommand, controllerPyFile.getAbsolutePath(), virtualEnvHome, listeningPort);
+        return processBuilder.start();
+    }
+
+
+    private void setupEnvironment() throws IOException {
+        final File environmentCreationCompleteFile = new File(virtualEnvHome, "env-creation-complete.txt");
+        if (environmentCreationCompleteFile.exists()) {
+            logger.debug("Environment has already been created for {}; will not recreate", virtualEnvHome);
+            return;
+        }
+
+        logger.info("Creating Python Virtual Environment {}", virtualEnvHome);
+
+        Files.createDirectories(virtualEnvHome.toPath());
+
+        final String pythonCommand = processConfig.getPythonCommand();
+        final String environmentPath = virtualEnvHome.getAbsolutePath();
+
+        final ProcessBuilder processBuilder = new ProcessBuilder(pythonCommand, "-m", "venv", environmentPath);
+        processBuilder.directory(virtualEnvHome.getParentFile());
+
+        final String command = String.join(" ", processBuilder.command());
+        logger.debug("Creating Python Virtual Environment {} using command {}", virtualEnvHome, command);
+        final Process process = processBuilder.start();
+
+        final int result;
+        try {
+            result = process.waitFor();
+        } catch (final InterruptedException e) {
+            throw new IOException("Interrupted while waiting for Python virtual environment to be created");
+        }
+
+        if (result != 0) {
+            throw new IOException("Failed to create Python Environment " + virtualEnvHome + ": process existed with code " + result);
+        }
+
+        // Create file so that we don't keep trying to recreate the virtual environment
+        environmentCreationCompleteFile.createNewFile();
+        logger.info("Successfully created Python Virtual Environment {}", virtualEnvHome);
+    }
```

Comment:
I can appreciate the intent. However, this class is really about managing the lifecycle of a Python process. To do this, it does require access to the specific member variables, etc. We could certainly abstract it away, but that would introduce additional complexity. And I don't believe there would ever be more than one implementation we would want to be used. While it would make it more testable, I would question the value of the tests we could develop. For example, would the tests be so specific that changing to the desired implementation means the tests are now failing and need to be changed? When we create a unit test, we want to create a test that verifies that the method under test functions properly and produces the expected output, given some inputs. I.e., that it adheres to the contract. But I feel anything that would be tested here would result in rather brittle tests. Do you see it differently?

Reply:
Good points, on further review, I can see how this would be difficult to abstract, and probably doesn't help in the end. As process management is key to the entire implementation, this will probably go through some iterations down the road, and could be revisited later.